File: baekjoon/9주차/2250-트리의-높이와-너비/우섭/2250.py
from sys import stdin

# 노드 구현
class Node:
    def __init__(self, number, left, right):
        self.parent = -1
        self.number = number
        self.left = left
        self.right = right

# 중위 탐색
def find(number, depth):
    global idx, maxdepth

    # 최대 깊이 저장
    if maxdepth < depth:
        maxdepth = depth

    # 왼쪽 노드 존재 시 탐색
    if tree[number].left != -1:
        find(tree[number].left, depth + 1)

    # 현재 노드를 기준으로
    # 동일한 깊이의 최소, 최대값 갱신
    minarr[depth] = min(minarr[depth], idx)
    maxarr[depth] = idx
    idx += 1

    # 오른쪽 노드 존재 시 탐색
    if tree[number].right != -1:
        find(tree[number].right, depth + 1)


n = int(stdin.readline())
tree = []
minarr = [n+1 for _ in range(n+1)]
maxarr = [0 for _ in range(n+1)]

# 기본 노트 세팅
for i in range(n+1):
    tree.append(Node(i, -1, -1))

# 노드 입력받기
for _ in range(n):
    num, l, r = map(int, stdin.readline().split(' '))
    tree[num].left = l
    tree[num].right = r
    tree[l].parent = num
    tree[r].parent = num

# 루트 노드 찾기
root = 0
for i in tree:
    if i.number != 0 and i.parent == -1:
        root = i.number
        break

# 탐색 진행
idx = 1
maxdepth = 1
find(root, 1)

# 넓이 최대값 및 해당 깊이 구하기
max = tmp = maxarr[1] - minarr[1] + 1
dep = 1
for i in range(1, maxdepth+1):
    tmp = maxarr[i] - minarr[i] + 1
    if tmp > max:
        max = tmp
        dep = i

print(f'{dep} {max}')

File: baekjoon/9주차/2250-트리의-높이와-너비/제문/2250.kt
package baekjoon

import java.lang.StringBuilder
import java.util.*
import kotlin.collections.HashMap

private val out = StringBuilder()
private val br = System.`in`.bufferedReader()

private fun readLine() = br.readLine()!!
private fun readInt() = br.readLine().toInt()
private fun readInts() = br.readLine().split(" ").map { it.toInt() }

private var count = 1
private val levels = HashMap<Int, MutableList<Int>>()

fun main() {
    val nodeN = readInt()
    val maps = HashMap<Int, MyTreeNode>()
    val tree = Tree()
    var answerIdx = 1
    var answer = 1

    repeat(nodeN) {
        val st = StringTokenizer(readLine(), " ")
        val element = st.nextToken().toInt()
        val left = st.nextToken().toInt()
        val right = st.nextToken().toInt()

        maps[element] = maps[element] ?: MyTreeNode(element, null, null, null)
        maps[left] = maps[left] ?: MyTreeNode(left, null, null, null)
        maps[right] = maps[right] ?: MyTreeNode(right, null, null, null)

        maps[element]!!.left = if (left != -1) maps[left] else null
        maps[element]!!.right = if (right != -1) maps[right] else null

        maps[left]?.parent = maps[element]
        maps[right]?.parent = maps[element]

        if (tree.head == null) tree.head = maps[element]
    }
    tree.findHead(tree.head)
    tree.inorder(tree.head, 1)

    for ((idx, key) in levels.keys.sorted().withIndex()) {
        val value = levels[key]!!.max()!! - levels[key]!!.min()!! + 1

        if (value > answer) {
            answer = value
            answerIdx = idx + 1
        }
    }
    print("$answerIdx $answer")
}

private class Tree {
    var head: MyTreeNode? = null

    fun inorder(curr: MyTreeNode?, depth: Int) {
        if (curr == null) return

        inorder(curr.left, depth + 1)
        levels[depth] = (levels[depth] ?: mutableListOf())
        levels[depth]!!.add(count++)
        inorder(curr.right, depth + 1)
    }

    fun findHead(curr: MyTreeNode?) {
        if (curr!!.parent == null) {
            head = curr
            return
        }

        findHead(curr.parent)
    }
}

private class MyTreeNode {
    var element: Int
    var left: MyTreeNode?
    var right: MyTreeNode?
    var parent: MyTreeNode?

    constructor(element: Int, left: MyTreeNode?, right: MyTreeNode?, parent: MyTreeNode?) {
        this.element = element
        this.left = left
        this.right = right
        this.parent = parent
    }
}

